tests/pos/i12739.scala

```diff
@@ -0,0 +1,15 @@
+object O {
+
+  class CA[A](var x: A)
+  type C = CA[_]
+
+  val c: CA[_] = CA(5)
+  def f[A](r: CA[A]): A = r.x
+
+  def g(): CA[_] = CA("hello")
+
+  f(g())
+  f(c)
+  f(c.asInstanceOf[CA[_]])
+  f(c.asInstanceOf[C])
+}
```

Comment:
Seeing as `f(c.asInstanceOf[c.type])` is the case that currently works in Scala 3 but doesn't in Scala 2, shall we add that one here too to make we track its state?